File: backend/node/api/src/application/use-cases/create-pipeline.ts
import { Pipeline } from "../../domain/entities/pipeline";
import { inject } from "../../infra/DI/container";
import type { PipelineRepository } from "../../infra/repository/pipeline";

type Input = {
	name?: string;
	description?: string;
	initialStepId?: string;
};

export class CreatePipelineUseCase {
	@inject("pipeline-repository")
	private declare readonly pipelineRepository: PipelineRepository;

	async execute(
		input: Input,
	): Promise<[string, undefined] | [undefined, Error]> {
		const [pipeline, error] = Pipeline.create(input);
		if (error) return [undefined, error];

		const [, saveError] = await this.pipelineRepository.save(pipeline);
		if (saveError) return [undefined, saveError];

		return [pipeline.getId(), undefined];
	}
}

File: backend/go/api/internal/pipeline/use-cases/create-pipeline.go
package usecases

import (
	"context"

	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/pipeline"
)

type CreatePipelineUseCaseDependencies struct {
	PipelineRepository *pipeline.PipelineRepository
}

type CreatePipelineUseCase struct {
	CreatePipelineUseCaseDependencies
}

func NewCreatePipelineUseCase(deps CreatePipelineUseCaseDependencies) CreatePipelineUseCase {
	return CreatePipelineUseCase{
		CreatePipelineUseCaseDependencies: deps,
	}
}

type CreatePipelineInput struct {
	Name        *string
	Description *string
	Canvas      *string
}

type CreatePipelineOutput struct {
	ID string
}

func (u *CreatePipelineUseCase) Execute(input CreatePipelineInput) (CreatePipelineOutput, error) {
	p, err := pipeline.NewPipeline(pipeline.NewPipelinePayload{
		Name:        input.Name,
		Description: input.Description,
		Canvas:      input.Canvas,
	})
	if err != nil {
		return CreatePipelineOutput{}, err
	}

	err = u.PipelineRepository.Save(context.Background(), p)
	if err != nil {
		return CreatePipelineOutput{}, err
	}

	return CreatePipelineOutput{ID: p.GetId()}, nil
}
